Enhance CAPI

oliverlloyd opened the pull request:
## What does this change?
Adds a new step to enhance the CAPI response 

![Screenshot 2020-05-03 at 11 51 44](https://user-images.githubusercontent.com/1336821/80912375-7a6cfc00-8d34-11ea-8c41-35e87700e3fb.jpg)

## What types of enhancements is this for?
- Drop caps (* * *) [[fontend code](https://github.com/guardian/frontend/blob/08b3732d113c28c5165ab2ce8e9b7b52909da336/common/app/views/support/HtmlCleaner.scala#L519)]
- Divider lines (also * * * but immersive only) [[fontend code](https://github.com/guardian/frontend/blob/08b3732d113c28c5165ab2ce8e9b7b52909da336/common/app/views/support/HtmlCleaner.scala#L539)]
- Auto linking [[fontend code](https://github.com/guardian/frontend/blob/08b3732d113c28c5165ab2ce8e9b7b52909da336/common/app/views/support/HtmlCleaner.scala#L849)?]
- Unicode bullets [[fontend code](https://github.com/guardian/frontend/blob/08b3732d113c28c5165ab2ce8e9b7b52909da336/common/app/views/support/HtmlCleaner.scala#L207)]
- Data attributes [[fontend code](https://github.com/guardian/frontend/blob/08b3732d113c28c5165ab2ce8e9b7b52909da336/common/app/views/support/HtmlCleaner.scala#L227)?]
- Photo essays [[frontend code](https://github.com/guardian/frontend/blob/08b3732d113c28c5165ab2ce8e9b7b52909da336/common/app/views/support/HtmlCleaner.scala#L453)]

## Why?
This initially helps us by giving us a way to replace some of the work done in [HtmlCleaner](https://github.com/guardian/frontend/blob/08b3732d113c28c5165ab2ce8e9b7b52909da336/common/app/views/support/HtmlCleaner.scala) on frontend. Having the code here, in DCR, speeds up development.

It also helps us identify and isolate these cleaning requirements so that we can work towards the goal of moving this logic further up the stack, away from the rendering tier.

PRBuilds commented:
PRbuilds results:










&#x1F49A; **AMP validation**
[amp-report.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-1438/amp-report.txt)







**LightHouse Reporting**
[1588503888.report.html](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-1438/1588503888.report.html)


--automated message

oliverlloyd commented:
![](https://github.trello.services/images/mini-trello-icon.png) [CAPI enhancement layer to enable html cleaning](https://trello.com/c/ngKRs7Bc/1585-capi-enhancement-layer-to-enable-html-cleaning)

shtukas commented:
Very nice! Also, the tests completely document what exactly any future upstream implementation must do. +1.

oliverlloyd commented:
@gtrufitt I refactored the enhance logic to match [your gist](https://gist.github.com/gtrufitt/1dc327066dbfbfa34a00ed61bd610549) but didn't quite make it all the way! Using a switch statement was a big improvement but I couldn't get the `(prev, current)` model to work, mainly because, you know, typescript. But I managed to get rid of the 'state' and things read a lot better now. Let me know what you think.

Oh, and I also renamed everything to be drop cap specific so we keep each 'enhancement' separate, with CAPIType contracts inbetween, so removing these later should be as simple as deleting one line and the file.

JamieB-gu commented:
This got me thinking... here's an immutable alternative that I _think_ should pass the tests:

```ts
const isText = (element: CAPIElement): boolean =>
    element !== undefined && element._type === TEXT;

const checkForDropCaps = (elements: CAPIElement[]): CAPIElement[] => {
    const [ element, ...tail ] = elements;

    if (element === undefined) {
        return [];
    }

    if (isDropCapFlag(element)) {
        const [ nextElement, ...tail2 ] = tail;

        if (isText(nextElement)) {
            return [ { ...nextElement, dropCap: true }, ...checkForDropcaps(tail2) ];
        }

        return checkForDropcaps(tail);
    }

    return [ element, ...checkForDropcaps(tail) ];
}
```

This may or may not be useful to you, I just thought it was an interesting problem 🙂

gtrufitt commented on src/web/server/render.ts:
```diff
@@ -3,12 +3,14 @@ import { extract as extractNAV } from '@root/src/model/extract-nav';
 export const render = ({ body }: express.Request, res: express.Response) => {
     try {
-        const CAPI = validateAsCAPIType(body);
+        const withDropCaps: CAPIType = addDropCaps(body);
```
I think maybe having a hop into a `enhanceCAPI` file, then then collates all the hacks might make it easier to reason with as render.ts could get large. But yeah, not urgent for this PR.

oliverlloyd replied:
I had the same thought 👍

gtrufitt commented on src/model/add-dropcaps.ts:
```diff
@@ -0,0 +1,68 @@
+import { JSDOM } from 'jsdom';
```
Yeah, think it's much cleaner to have each 'mod' in a separate file. Does 'model' describe the folder well? Maybe `enhanceCAPI` or `hacks` - model feels a little too generic and I think it's worth have the folder name descriptive.

oliverlloyd replied:
Maybe if we call it `hacks` it's a clear message that we want to remove this code at a later stage

gtrufitt commented on src/model/add-dropcaps.ts:
```diff
@@ -0,0 +1,68 @@
+import { JSDOM } from 'jsdom';
+
+const TEXT = 'model.dotcomrendering.pageElements.TextBlockElement';
```
Maybe just inline these?

oliverlloyd commented:
> This got me thinking... here's an immutable alternative that I _think_ should pass the tests:
> 
> ```ts
> const isText = (element: CAPIElement): boolean =>
>     element !== undefined && element._type === TEXT;
> 
> const checkForDropCaps = (elements: CAPIElement[]): CAPIElement[] => {
>     const [ element, ...tail ] = elements;
> 
>     // End of the list
>     if (element === undefined) {
>         return [];
>     }
> 
>     if (isDropCapFlag(element)) {
>         const [ nextElement, ...tail2 ] = tail;
> 
>         // If TEXT, add the `dropCap` flag
>         if (isText(nextElement)) {
>             return [ { ...nextElement, dropCap: true }, ...checkForDropcaps(tail2) ];
>         }
> 
>         // Otherwise remove the dropcap and continue
>         return checkForDropcaps(tail);
>     }
> 
>     // No dropcaps here, so append the element and continue
>     return [ element, ...checkForDropcaps(tail) ];
> }
> ```
> 
> This may or may not be useful to you, I just thought it was an interesting problem 🙂

Using the `rest` trick when spreading is a clever way to go about this!

My instinct is to keep things a bit more basic (albeit more verbose) though - a lot of people don't know how to do this and it can be hard to reason about when you're not used to it
